src/compiler/checker.ts

```diff
@@ -12327,11 +12315,24 @@ namespace ts {
                 }
                 else {
                     if (target.flags & TypeFlags.Union) {
-                        result = typeRelatedToSomeType(source, <UnionType>target, reportErrors && !(source.flags & TypeFlags.Primitive) && !(target.flags & TypeFlags.Primitive));
+                        result = typeRelatedToSomeType(getRegularTypeOfObjectLiteral(source), <UnionType>target, reportErrors && !(source.flags & TypeFlags.Primitive) && !(target.flags & TypeFlags.Primitive));
+                        if (result && isPerformingExcessPropertyChecks) {
+                            // Validate against excess props using the original `source`
+                            const discriminantType = target.flags & TypeFlags.Union ? findMatchingDiscriminantType(source, target as UnionType) : undefined;
+                            if (!propertiesRelatedTo(source, discriminantType || target, reportErrors)) {
+                                return Ternary.False;
+                            }
+                        }
                     }
                     else if (target.flags & TypeFlags.Intersection) {
                         isIntersectionConstituent = true; // set here to affect the following trio of checks
-                        result = typeRelatedToEachType(source, target as IntersectionType, reportErrors);
+                        result = typeRelatedToEachType(getRegularTypeOfObjectLiteral(source), target as IntersectionType, reportErrors);
+                        if (result && isPerformingExcessPropertyChecks) {
+                            // Validate against excess props using the original `source`
+                            if (!propertiesRelatedTo(source, target, reportErrors)) {

```

Comment:
Is the `propertiesRelatedTo` check more expensive than the `typeRelatedToEachType` check? Does it make sense to do the checks in the other order?

Reply:
I think usually, yeah - we typically consider structural decomposition more expensive than algebraic comparison.

Reply:
Ah ok, fair enough. I thought that `typeRelatedToEachType` would end up calling into `propertiesRelatedTo` for each union component anyway via `isRelatedTo`.

Reply:
It can; but that's the last kind of comparison we make - any other comparison succeeding avoids it.

Reply:
That makes sense. Ty!